### PR TITLE
Do not use Ktor in Js HttpEngine, use fetch directly instead

### PR DESCRIPTION
**Pull request by BoD**

### CODE REVIEW

**Review by martinbonnin — APPROVED** [review]
A few questions but overall very nice 👍 👍

File: tests/engine/src/commonTest/kotlin/HttpEngineTest.kt [hunk 1]
```diff
@@ -104,10 +97,7 @@ class HttpEngineTest {
             .build()
     )
     assertFailsWith<ApolloException> {
-      httpEngine.execute(HttpRequest.Builder(HttpMethod.Get, mockServer.url()).build())
+      httpEngine(500).execute(HttpRequest.Builder(HttpMethod.Get, mockServer.url()).build())
     }
   }
-
-  @Test
-  fun readTimeout() = readTimeout(httpEngine(500))
```
[the remark this answers is not in the document]
Comment (BoD) [reply tests/engine/src/commonTest/kotlin/HttpEngineTest.kt:112]: Exactly ✅

File: libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt [hunk 2]
```diff
@@ -1,85 +1,156 @@
...
+  @Suppress("UnsafeCastFromDynamic")
+  override suspend fun execute(request: HttpRequest): HttpResponse {
+    val abortController = AbortController()
+    val connectTimeoutId = setTimeout({ abortController.abort() }, connectTimeoutMillis)
 
-      // socketTimeoutMillis would make more sense but doesn't seem to work on JS. See https://youtrack.jetbrains.com/issue/KTOR-6211
-      this.requestTimeoutMillis = this@JsHttpEngine.readTimeoutMillis
+    val fetchOptions = request.toFetchOptions(abortSignal = abortController.signal)
+    val responsePromise: Promise<Response> = if (isNode) {
+      requireNodeFetch()(
```
[the remark this answers is not in the document]
Comment (BoD) [reply libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt:45]: We can probably call `requireNodeFetch()` only once 👀 , let me check.

File: libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt [hunk 3]
```diff
@@ -1,85 +1,156 @@
...
-          it.writeTo(buffer)
-          setBody(buffer.readUtf8())
-        }
+    return try {
+      val response = responsePromise.await()
+      clearTimeout(connectTimeoutId)
+      val responseBodySource = if (isNode) {
+        readBodyNode(response.body, readTimeoutMillis, abortController)
+      } else {
+        readBodyBrowser(response.body, readTimeoutMillis, abortController)
```
[the remark this answers is not in the document]
Comment (BoD) [reply libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt:61]: From what I saw, `body` from the w3c's fetch vs the Node one are not exactly the same API, but I actually haven't tried to use `.read` on the Node one, maybe it works too (🪄 ) - I'll try it.

Also, with Node v21, we'll be able to just use the regular w3c `fetch` without the need to import the node-fetch library - but for now we depend on v18.

File: libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (BoD) [reply libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt]: Done in 669614877b7fe952b674430f5189971e429845fb

File: libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (BoD) [reply libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt:63]: Well `read` doesn't seem to work on node.

[the remark this answers is not in the document]
Comment (martinbonnin) [reply libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.js.kt:63]: Alright then 👍

**Review by martinbonnin — APPROVED** [review]
LGTM but marking this as draft in order to land other PRs before this one for 4.0.0-beta.5

**Comment by martinbonnin** [comment]
4.0.0-beta.5 is released, feel free to merge

**Pull request merged**